### PR TITLE
feat: Add idempotency feature to detect duplicate requests due to network conditions

### PR DESCRIPTION
**Pull request by DrMimik**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Android/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-Android/issues/764).

### Issue Description
Requests are duplicated on slow internet connections and during network interruptions.

Closes: [#764](https://github.com/parse-community/Parse-SDK-Android/issues/764)

### Approach
Add idempotency requestId header for client requests to enforce server side idempotency middleware on selected routes..

### TODOs before merging

- [x] Add tests
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)

### CODE REVIEW

File: parse/src/main/java/com/parse/ParseRESTCommand.java [hunk 1]
```diff
@@ -215,6 +216,20 @@ protected void addAdditionalHeaders(ParseHttpRequest.Builder requestBuilder) {
...
+            // using header names so we don't override a parameter with the same key name,
+            // using all headers to insure the requestId generated doesn't conflict with the rest of the users
+            if (installationId != null)
+                jsonObject.put(HEADER_INSTALLATION_ID, installationId);
+            if (sessionToken != null)
+                jsonObject.put(HEADER_SESSION_TOKEN, sessionToken);
+            if (masterKey != null)
+                jsonObject.put(HEADER_MASTER_KEY, masterKey);
+            requestBuilder.addHeader(HEADER_REQUEST_ID, ParseDigestUtils.md5(toDeterministicString(jsonObject)));
+        } catch (JSONException e) {
```
Comment (mtrezza) [inline parse/src/main/java/com/parse/ParseRESTCommand.java:230]: The OS should provide a UUID API that takes care of seeding. Could you use the same logic for generating the UUID as we use in the other SDKs? For example the iOS SDK and JS SDK?

File: parse/src/main/java/com/parse/ParseRESTCommand.java [hunk 2]
```diff
@@ -35,9 +35,10 @@ class ParseRESTCommand extends ParseRequest<JSONObject> {
     /* package */ static final String HEADER_OS_VERSION = "X-Parse-OS-Version";
 
     /* package */ static final String HEADER_INSTALLATION_ID = "X-Parse-Installation-Id";
+    /* package */ static final String HEADER_REQUEST_ID = "X-Parse-Request-Id";
     /* package */ static final String USER_AGENT = "User-Agent";
-    private static final String HEADER_SESSION_TOKEN = "X-Parse-Session-Token";
-    private static final String HEADER_MASTER_KEY = "X-Parse-Master-Key";
```
Comment (mtrezza) [inline parse/src/main/java/com/parse/ParseRESTCommand.java:40]: Revert, shouldn't be needed

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: parse/src/main/java/com/parse/ParseRESTCommand.java (hunk 1 above)
Comment (DrMimik) [reply parse/src/main/java/com/parse/ParseRESTCommand.java]: I have rewritten the logic to use a UUID, as this is uniformed between SDKs.

But imo why isn't the UUID data based?, like this if the SDK calls ParseObject.save() it changes the ParseRequest which creates a new UUID, while using a data based UUID calling save won't change the UUID header.

Currently if I want the SDK to create an idempotent request I'd still set the HttpClient to use a longer timeout and increase the default max retries of requests.

**Comment by DrMimik** [comment]
Should I force push the first commit? or use a new commit?.

File: parse/src/main/java/com/parse/ParseRESTCommand.java (hunk 1 above)
Comment (DrMimik) [reply parse/src/main/java/com/parse/ParseRESTCommand.java]: I am not very good at JS, but could you direct me to where the iOS SDK impliments the logic.

Comment (mtrezza) [reply parse/src/main/java/com/parse/ParseRESTCommand.java]: > But imo why isn't the UUID data based?

It doesn't need to be based on object data. A UUID is a Universally **Unique** IDentifier, which ensures that every generated UUID is unique.

The part you'd have to take care about in the PR is that for a `ParseObject.save` that UUID is only generated one time and then reused if the request fails and the SDK auto-retries the request. Obviously, if you generated a new UUID with ever retry of the same request, then this wouldn't work.

Comment (DrMimik) [reply parse/src/main/java/com/parse/ParseRESTCommand.java]: The current approach works as follows if all requests were unsuccessfull:

`ParseObject object = new ParseObject("GameScore");`
// calls a request and retries according to the max retries (default is 4) set during initilization, with the same requestId header. (5 calls with the same header)
`object.save();`

// this second method call... calls a request and retries according to the max retries set, with the same requestId header but different than the first save() method call.
`object.save();`

Comment (mtrezza) [reply parse/src/main/java/com/parse/ParseRESTCommand.java]: The second `save` call should use a different UUID for the request ID. In fact, every `save` call should use its own UUID. Only auto-retries of each `save` call should use the same UUID.

Comment (DrMimik) [reply parse/src/main/java/com/parse/ParseRESTCommand.java]: Yes, It is using exactly that. My explanation was kind of confusing.

I'll add a second test using the code in my previous comment.

**Comment by azlekov** [comment]
> Should I force push the first commit? or use a new commit?.

New commit is preferable to track changes and reviewing it easily.
Hoping to get this merged.

**Comment by dplewis** [comment]
@mtrezza Is this ready to merge?

**Review by mtrezza — CHANGES_REQUESTED** [review]
Lint failed

**Comment by mtrezza** [comment]
@parse-community/android-sdk anyone wants to pick this up and submit a new PR?

**Comment by abubkr-hago** [comment]
Hi, I am a new contributor, I would love to continue on this PR

**Comment by mtrezza** [comment]
Sure, please go ahead.

**Pull request closed without merging**